### PR TITLE
Handle list() errors more gracefully

### PR DESCRIPTION
**Pull request by mcronce**
Currently, running both with my own storage backend as well as with `new_with_fs_root()`, when an error occurs in `list()`, the server doesn't send a reply at all - from the FTP client's standpoint, it just hangs.  Running with logging turned up it does print something to the log, which is a plus, but hanging the FTP session with something as simple as `ls /path/that/does/not/exist` doesn't seem ideal :)

With the changes in this PR, I bubble up any errors that occurred in `list()` for logging purposes, and send a reply to the client indicating failure.  Additionally, the rest of the steps occur the same as they do on success - closing the data socket, then notifying the main thread (although the internal message differs) that we're finished.

Related to #163 in spirit, but probably not in letter...suggestions welcome.  If this looks good it may make sense to apply the same treatment to other storage backend wrappers.

### CODE REVIEW

**Comment by hannesdejager** [comment]
Hi @mcronce , thanks for taking the time. Good one!

Before:

```
lftp localhost:~> ls /path/that/does/not/exist
`ls /path/that/does/not/exist' at 0 [Waiting for response...]
```
After:

```
lftp localhost:~> ls /path/that/does/not/exist
550 Permanent file not available
```

I'm curious as to what you're own storage back-end use-case is and what your experience was like implementing it. It is always nice to hear from people using what is put out there. If you are able to share please do.

**Comment by mcronce** [comment]
Definitely.  It's not open-source yet, but will be pretty soon, as the project is reaching a state where I consider it presentable :) I'll link it when it's published.

I'm making a document store (along the lines of Paperless, Mayan EDMS, etc etc etc...I tried at least a half dozen projects and none of them met all my requirements) and one of the features I wanted was FTP and/or CIFS and/or NFS interfaces so that (A) my scanner could upload directly to it, rather than having a watch directory and a cronjob, and (B) I could mount it on my desktop/laptop to bulk upload stuff I've been saving in various places over the years because no web interface I'm going to make will be better than `cp -r` for that purpose.

The one particularly weird thing I'm doing is that I needed two top-level trees - one for the actual directory tree of documents, and one for the tag hierarchy, exposing the documents in each tag as a symlink (which I don't have fully working yet, but it's getting there).

After finding libunftp, I decided FTP would definitely come first, because I wouldn't have to implement the line protocol myself in Rust ;)

My experience writing my own storage backend has been very very positive.  It's pretty much _exactly_ what I would want.  Implement the handful of what I would consider "fundamental" filesystem operations, instantiate the library-provided server, and boom, working FTP server.  No extra hoops; I don't even really need to think about FTP specifically.  In fact, when I go to do CIFS and NFS at some future date, it is very likely that I will model them after libunftp - implement the protocol, a server, a trait, and a reference implementation of the trait.

tl;dr Awesome lib, do recommend.

**Comment by mcronce** [comment]
https://gitlab.cronce.io/yadms/yadms-ftp-server is the FTP service, it needs a lot of cleanup still but it's mostly functional

**Comment by hannesdejager** [comment]
Ah, thanks for sharing and also the code, great to see.

I've just pushed version 0.11.0 to crates.io so you can remove your versioning TODO :-). There were some breaking changes but I don't think it will impact you.

**Comment by mcronce** [comment]
It doesn't look it did.  Thanks for the heads up :)
